### PR TITLE
pkg/histogram: Create new package for all gadgets using histograms

### PR DESCRIPTION
**Pull request by blanquicet**
This PR moves the functions that allow generating and printing a histogram from the profile block-io gadget to an internal package. It also adds unit-tests for such package.

This new package will be used also by the upcoming gadget: [profile tcprtt](https://github.com/inspektor-gadget/inspektor-gadget/pull/1517).

### CODE REVIEW

File: pkg/gadgets/internal/histogram/histogram.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	Slots []Slot `json:"slots,omitempty"`
+}
+
+// New creates a new Histogram.
+func New(unit string, slots []Slot) *Histogram {
+	return &Histogram{
+		Unit:  unit,
+		Slots: slots,
+	}
+}
```
Comment (mauriciovasquezbernal) [inline pkg/gadgets/internal/histogram/histogram.go:47]: Do we really need this function if Histogram is an exported type?

File: pkg/gadgets/internal/histogram/histogram_test.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+			description: "Empty histogram",
+			histogram:   New("usecs", []Slot{}),
+			expected:    "",
+		},
+		{
+			description: "With 1 slot value 1",
+			histogram:   NewFromCSlots("usecs", []uint32{1}),
+			expected: "" +
+				"     usecs               : count    distribution\n" +
+				"         1 -> 1          : 1        |****************************************|\n",
```
[the remark this answers is not in the document]
Comment (blanquicet) [reply pkg/gadgets/internal/histogram/histogram_test.go:123]: I don't know. I know the concatenation of string doesn't look that great, but IMO, it makes things more organized and easy to create. It took me a while to match the spaces at the beginning of the line when using the backtick. I prefer keeping the current implementation 🙈.

File: pkg/gadgets/internal/histogram/histogram.go (hunk 1 above)
Comment (blanquicet) [reply pkg/gadgets/internal/histogram/histogram.go:47]: Not really, but it looked weird to me to have `NewFromCSlots` and not `New` 😅. I'm ok removing it.

File: pkg/gadgets/internal/histogram/histogram.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	IntervalStart uint64 `json:"intervalStart"`
+	IntervalEnd   uint64 `json:"intervalEnd"`
+}
+
+// Histogram represents a histogram of the number of events that occurred in
+// each interval.
+type Histogram struct {
+	Unit  string `json:"unit,omitempty"`
+	Slots []Slot `json:"slots,omitempty"`
+}
```
Comment (blanquicet) [inline pkg/gadgets/internal/histogram/histogram.go:39]: I considered defining a `Unit` type with its possible values like this:

```suggestion
type Unit string

const (
	UnitMilliseconds Unit = "msecs"
	UnitMicroseconds Unit = "usecs"
)

// Histogram represents a histogram of the number of events that occurred in
// each interval.
type Histogram struct {
	Unit  Unit   `json:"unit,omitempty"`
	Slots []Slot `json:"slots,omitempty"`
}
```

It will simplify the caller's life by allowing this:

```golang
h := histogram.NewFromCSlots(histogram.UnitMicroseconds, []uint32{1}),
```

Instead of:

```golang
h := histogram.NewFromCSlots("usecs", []uint32{1}),
```

However, it'll require exporting this package in `pkg` to make it usable also by integration tests. WDYT?

cc @eiffel-fl @mauriciovasquezbernal

Comment (blanquicet) [reply pkg/gadgets/internal/histogram/histogram.go:39]: > I was wandering about the unit but so far it seems to only be printed. I mean, if we change the unit it does not have any impact on what is drawn (i.e. the underlying computation will not be done in another unit than µs), right?
>
The profile block-io gadget only supports microseconds, so this change is just a matter of using a variable instead of a literal string. Please notice that the bpf code is able to provide info in micro and milliseconds, but we are not using such a flag for this gadget.

> It could nonetheless be a good idea to introduce this new type as it would ease adding a PR to make the whole configurable if we want to use other time.
>
Exactly. I added support for micro and milliseconds in the profile tcprtt gadget that I'm working so this new type would help (also when we will add the same support to profile block-io).

> Nonetheless, rather than declaring our own type, would it be possible to use something from time golang package?
> 
I didn't find anything. The time package uses another notation: `µs`, and don't use constants, check [this](https://cs.opensource.google/go/go/+/refs/tags/go1.20.3:src/time/time.go;l=644). Do you know any package that provides it?

File: pkg/gadgets/internal/histogram/histogram.go (hunk 1 above)
Comment (mauriciovasquezbernal) [reply pkg/gadgets/internal/histogram/histogram.go]: A possibility is to make `parseCSlots` -> `NewSlots()` and remove `New` and `NewFromCSlots`. Then the user could create a histogram as:

```
myHist := &Histogram{
		Unit:  unit,
		Slots: NewSlots(cSlots),
	}
```

File: pkg/histogram/histogram.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+import (
+	"fmt"
+	"strings"
+)
+
+type Unit string
+
+const (
+	UnitMilliseconds Unit = "msecs"
+	UnitMicroseconds Unit = "usecs"
```
[the remark this answers is not in the document]
Comment (blanquicet) [reply pkg/histogram/histogram.go:32]: I'm not sure, "msecs" and "usecs" comes from BCC. But it is also true that we created a generic histogram representation at this point. I'm ok either way.

File: pkg/histogram/histogram.go [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+		parsedSlots = append(parsedSlots, Slot{
+			Count:         uint64(val),
+			IntervalStart: start,
+			IntervalEnd:   end,
+		})
+	}
+
+	// Slots are 0-indexed, so we need to increment indexMax to get the number
+	// of slots.
```
Comment (mauriciovasquezbernal) [inline pkg/histogram/histogram.go:76]: I'm not sure about this comment. I think this is because the expression `s[a:b]` will take the elements `s[a]`, ... , `s[b-1]`, so if we want to take up to the `s[x]` element we have to use `s[:x+1]`.  (Somehow this is related to the fact that arrays are 0-indexed but the comment is not straightforward for me)

File: pkg/histogram/histogram.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/histogram/histogram.go:32]: IMO it's better to deviate from BCC and keep it aligned with other time representations that we have and use `time.Duration`

File: pkg/histogram/histogram.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+	parsedSlots := make([]Slot, 0, len(cSlots))
+	indexMax := 0
+	for i, val := range cSlots {
+		if val > 0 {
+			indexMax = i
+		}
+
+		start := uint64(1) << i
+		end := (uint64(1) << (i + 1)) - 1
```
Comment (mauriciovasquezbernal) [inline pkg/histogram/histogram.go:63]:
```suggestion
		start := uint64(1) << i
		end := 2*start - 1
```

Perhaps that helps a bit with code readability?

File: pkg/histogram/histogram.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+)
+
+type Unit string
+
+const (
+	UnitMilliseconds Unit = "msecs"
+	UnitMicroseconds Unit = "usecs"
+)
+
+type Slot struct {
```
Comment (mauriciovasquezbernal) [inline pkg/histogram/histogram.go:35]: nit: I don't know if `Slot` is the better terminology here. It seems that `Bucket` is the standard. I also think `Interval` can work.

File: pkg/histogram/histogram.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package histogram provides a Histogram struct that represents a histogram of
+// the number of events that occurred in each interval. The intervals are powers
+// of two, so the slots array is indexed by the power of two minus one. For
+// example, the first slot is the number of events that occurred in the interval
+// [1, 1], the second slot is [2, 3], the third slot is [4, 7], and so on.
```
Comment (mauriciovasquezbernal) [inline pkg/histogram/histogram.go:19]: `[1, 1]` -> `[0, 1]` ? :exploding_head:

File: pkg/histogram/histogram.go [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+	if valMax == 0 {
+		return strings.Repeat(" ", int(width))
+	}
+
+	minVal := uint64(0)
+	if val < valMax {
+		minVal = val
+	} else {
+		minVal = valMax
+	}
```
Comment (mauriciovasquezbernal) [inline pkg/histogram/histogram.go:129]: Is this logic needed given that there is some logic to determine valMax in L88?

File: pkg/histogram/histogram.go [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+	stars := minVal * width / valMax
+	spaces := width - stars
+
+	var sb strings.Builder
+	sb.WriteString(strings.Repeat("*", int(stars)))
+	sb.WriteString(strings.Repeat(" ", int(spaces)))
+	if val > valMax {
+		sb.WriteByte('+')
+	}
```
Comment (mauriciovasquezbernal) [inline pkg/histogram/histogram.go:139]: ditto.

File: pkg/histogram/histogram.go [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+	valMax := uint64(0)
+	for _, slot := range h.Slots {
+		if slot.Count > valMax {
+			valMax = slot.Count
+		}
+	}
+
+	// reportEntries maximum value is C.MAX_SLOTS which is 27, so we take the
+	// value when idx_max <= 32.
```
Comment (mauriciovasquezbernal) [inline pkg/histogram/histogram.go:96]: Is this a comment from the original code that doesn't make sense now?

File: pkg/histogram/histogram.go (hunk 8 above)
Comment (mauriciovasquezbernal) [inline pkg/histogram/histogram.go:19]: I think this histogram struct can be generic (linear, exp2, etc). This comment is more related to the `NewSlots` function.

File: pkg/histogram/histogram.go [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+// Histogram represents a histogram of the number of events that occurred in
+// each interval.
+type Histogram struct {
+	Unit  Unit   `json:"unit,omitempty"`
+	Slots []Slot `json:"slots,omitempty"`
+}
+
+// NewSlots creates a new Slot array from a C-Slots array returned by the BPF
+// program.
+func NewSlots(cSlots []uint32) []Slot {
```
Comment (mauriciovasquezbernal) [inline pkg/histogram/histogram.go:50]: As mentioned above, I think we need to make it clear that the `cSlots` correspond to a exp2 histogram. (We calculate start and end based on that assumption). Perhaps this can be renamed to `NewExp2Slots` ?

**Review by mauriciovasquezbernal — COMMENTED** [review]
I didn't intended to review this into too much detail, but once I started I couldn't stop. Sorry if I added some many detail comments, I know your intention was only to move some code, but we can take the opportunity to improve it.

File: pkg/histogram/histogram.go (hunk 5 above)
Comment (blanquicet) [reply pkg/histogram/histogram.go:76]: Yeah, the comment is not clear enough:

```suggestion
	// The element parsedSlots[indexMax] is the last element with a non-zero
	// value. So, we need to use parsedSlots[:indexMax+1] to include it in the
	// returned array.
```

File: pkg/histogram/histogram.go [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+	for i, val := range cSlots {
+		if val > 0 {
+			indexMax = i
+		}
+
+		start := uint64(1) << i
+		end := (uint64(1) << (i + 1)) - 1
+		if start == end {
+			start -= 1
+		}
```
[the remark this answers is not in the document]
Comment (blanquicet) [reply pkg/histogram/histogram.go:66]: Yep, it is only valid in that case.

File: pkg/histogram/histogram.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (blanquicet) [reply pkg/histogram/histogram.go:32]: It is fine for me. Just one thing. Should we use "us" or "µs"? The time package uses the Greek letter.

File: pkg/histogram/histogram.go (hunk 7 above)
Comment (blanquicet) [reply pkg/histogram/histogram.go:35]: Yeah, [bucket](https://en.wikipedia.org/wiki/Data_binning) seems to be the standard.

File: pkg/histogram/histogram.go (hunk 5 above)
Comment (mauriciovasquezbernal) [reply pkg/histogram/histogram.go:76]: Yes, it is.

File: pkg/histogram/histogram.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/histogram/histogram.go:32]: `µs` should be fine.

File: pkg/histogram/histogram.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (mauriciovasquezbernal) [reply pkg/histogram/histogram.go:66]: Should it be `start = 0` on your code?

[the remark this answers is not in the document]
Comment (blanquicet) [reply pkg/histogram/histogram.go:66]: Yep, fixed.

File: pkg/histogram/histogram.go [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package histogram provides a Histogram struct that represents a histogram of
+// the number of events that occurred in each interval. The intervals are powers
+// of two, so the slots array is indexed by the power of two minus one. For
+// example, the first slot is the number of events that occurred in the interval
+// [1, 1], the second slot is [2, 3], the third slot is [4, 7], and so on.
+
```
[the remark this answers is not in the document]
Comment (blanquicet) [reply pkg/histogram/histogram.go:20]: It is not required so I removed it to avoid problems.

File: pkg/histogram/histogram.go (hunk 11 above)
Comment (blanquicet) [reply pkg/histogram/histogram.go:96]: At least for me, it doesn't make sense now.

File: pkg/histogram/histogram.go (hunk 9 above)
Comment (blanquicet) [reply pkg/histogram/histogram.go:129]: Yep, IMO, the computation of `stars` can be done directly with `val` as `val` will never be greater than `valMax`. It means `stars := val * width / valMax`.

File: pkg/histogram/histogram.go (hunk 10 above)
Comment (blanquicet) [reply pkg/histogram/histogram.go:139]: Same, `val` will never be greater than `valMax`. Removed.

File: pkg/histogram/histogram_test.go [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+		expected    *Histogram
+	}{
+		{
+			description: "Empty slots",
+			slots:       []uint32{},
+			expected: &Histogram{
+				Unit:      unit,
+				Intervals: nil,
+			},
+		},
```
Comment (mauriciovasquezbernal) [inline pkg/histogram/histogram_test.go:40]: nit: add a test with `slots: nil` to be sure it doesn't crash.

File: pkg/histogram/histogram_test.go [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+		{
+			description: "With 1 slot",
+			slots:       []uint32{1},
```
Comment (mauriciovasquezbernal) [inline pkg/histogram/histogram_test.go:43]: nit: I'd use higher numbers than 1 in this case, to avoid those being messed up with indexes. (8888, 9999, ...)

**Review by mauriciovasquezbernal — APPROVED** [review]
LGTM!
